geomstats/geometry/functions.py

```diff
@@ -0,0 +1,245 @@
+        """Compute the Riemannian exponential of a tangent vector.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., n_evals]
+            Tangent vector at a base point.
+        base_point : array-like, shape=[..., n_evals]
+            Point on the hypersphere.
+
+        Returns
+        -------
+        exp : array-like, shape=[..., n_evals]
+            Point on the hypersphere equal to the Riemannian exponential
+            of tangent_vec at the base point.
+        """
+
+        def _f(v, p):
+            norm_v = self.norm(v)
+            t1 = gs.cos(norm_v) * p
+            t2 = (gs.sin(norm_v) / norm_v) * p
+            return t1 + t2
+
+        out = gs.vectorize(
+            (tangent_vec, base_point),
+            _f,
+            dtype=gs.float32,
+            multiple_args=True,
+            signature="(i),(i)->(i)",
+        )
+
+        return out
+
+    def log(self, point, base_point, **kwargs):
+        """Compute the Riemannian logarithm of a point.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., n_evals]
+            Point on the hypersphere.
+        base_point : array-like, shape=[..., n_evals]
+            Point on the hypersphere.
+
+        Returns
+        -------
+        log : array-like, shape=[..., n_evals]
+            Tangent vector at the base point equal to the Riemannian logarithm
+            of point at the base point.
+        """
+
+        def _f(p0, p1):
+            theta = gs.arccos(self.inner_product(p1, p0, base_point=None))
+            return (p1 - p0 * gs.cos(theta)) * (theta / gs.sin(theta))
+
+        out = gs.vectorize(
+            (base_point, point),
+            _f,
+            dtype=gs.float32,
+            multiple_args=True,
+            signature="(i),(i)->(i)",
+        )
+
+        return out
+
+
+class SinfSpace(Manifold):
+    """Class for space of L2 functions with norm 1.
+
+    The tangent space is given by functions that have
+    zero inner-product with the base point
+
+    Parameters:
+    -------
+    domain_samples : array of shape (n_samples, )
+        grid points on the domain
+
+    Ref :
+    -----
+    .. [Sea2016] Srivastava, Anuj, and Eric P. Klassen.
+    Functional and shape data analysis.
+    Vol. 1. New York: Springer, 2016.
+    """
+
+    def __init__(self, domain_samples):
+        self.domain = domain_samples
+        super().__init__(dim=math.inf, metric=SinfSpaceMetric(self.domain))
+
+    def projection(self, point):
+        """Project a point to the infinite dimensional hypersphere.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., dim]
+            discrete evaluation of a function.
+        Returns
+        -------
+        projected_point : array-like, shape=[..., dim]
+            Point projected to the hypersphere.
+        """
+        norm_p = self.metric.norm(point)
+        return point / norm_p
```

Comment:
Does this work for multiple samples?

Reply:
I think the updated one should work! thanks for pointing this out.